lib/core/src/configuration_hook_manager.cpp

```diff
@@ -0,0 +1,47 @@
+#include "irods/configuration_hook_manager.hpp"
+#include "irods/irods_logger.hpp"
+
+#include <unordered_map>
+
+namespace irods::experimental::configuration_reload
+{
+    namespace
+    {
+        // This is where the function objects are stored, given that they will not be removed or reordered, their index
+        // is what edge_sets' value refers to.
+        std::vector<hook> hooks;
+
+        // Edge set, as in edge-triggered, contains an association between the changed property and an index into hooks.
+        std::unordered_multimap<std::string, size_t> edge_sets;
+    } // namespace
+
+    void add_config_reload_hook(const auto& hook_triggers, hook&& hook)
+    {
+        size_t id = hooks.size();
+        // Take ownership, we don't need to deal with any missing closures.
+        // those can be nasty to debug.
+        hooks.emplace_back(std::move(hook));
+        for (const auto& name : hook_triggers) {
+            edge_sets.insert(name, id);
+        }
+    }
+
+    void run_hooks(const nlohmann::json& changes)
+    {
+        // The *CATCHALL* hook is triggered on every change in configuration value.
+        auto catch_all_hooks = edge_sets.equal_range("*CATCHALL*");
```

Comment:
I'm still very confused by the use of `"*CATCHALL*"`.

Is it used somewhere else?
Is this something the caller is supposed to know about?

I don't see any other references to it in the PR.